staging/src/k8s.io/externaljwt/pkg/plugin/keycache.go

```diff
@@ -0,0 +1,203 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+	"k8s.io/klog/v2"
+)
+
+// Listener is an interface to use to notify interested parties of a change.
+type Listener interface {
+	// Enqueue should be called when an input may have changed
+	Enqueue()
+}
+
+type KeyCache struct {
+	plugin *Plugin
+
+	syncGroup singleflight.Group
+
+	publicKeysLock sync.RWMutex
+	publicKeys     *VerificationKeys
+
+	listenersLock sync.Mutex
+	listeners     []Listener
+
+	ticker             *time.Ticker
+	refreshIntervalSec int
+}
+
+// newKeyCache constructs an implementation of KeyCache that wraps external JWT signer plugin.
+func newKeyCache(plugin *Plugin) *KeyCache {
+	cache := &KeyCache{
+		plugin:             plugin,
+		refreshIntervalSec: 600, // Default refresh interval is 10 mins
+	}
+	return cache
+}
+
+// InitialFill can be used to performs an initial fetch for keys get the
+// refresh interval as recommended by external signer.
+func (p *KeyCache) InitialFill(ctx context.Context) error {
+	if err := p.syncKeys(ctx); err != nil {
+		return fmt.Errorf("while performing initial cache fill: %w", err)
+	}
+	return nil
+}
+
+func (p *KeyCache) StartPeriodicSync(ctx context.Context) {
+	p.ticker = time.NewTicker(time.Second * time.Duration(p.refreshIntervalSec))
+	defer p.ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			klog.InfoS("Key cache shutting down")
+			return
+		case <-p.ticker.C:
+		}
+
+		if err := p.syncKeys(ctx); err != nil {
+			klog.Errorf("when syncing supported public keys(Stale set of keys will be supported): %v", err)
+			continue
+		}
+
+		p.broadcastUpdate()
+	}
+}
+
+func (p *KeyCache) AddListener(listener Listener) {
+	p.listenersLock.Lock()
+	defer p.listenersLock.Unlock()
+
+	p.listeners = append(p.listeners, listener)
+}
+
+func (p *KeyCache) GetCacheAgeMaxSeconds() int {
+	// consumer shall support the keys for no longer than 10 mins before refreshing.
+	return p.refreshIntervalSec
+}
+
+// GetPublicKeys returns the public key corresponding to requested keyID.
+// Getter is expected to return All keys for keyID ""
+func (p *KeyCache) GetPublicKeys(keyID string) []PublicKey {
+	ctx := context.Background()
+
+	pubKeys, ok := p.findKeyForKeyID(keyID)
+	if ok {
+		return pubKeys
+	}
+
+	// If we didn't find it, trigger a sync.
+	if err := p.syncKeys(ctx); err != nil {
+		klog.ErrorS(err, "Error while syncing keys")
```

Comment:
Return early?

Reply:
You mean we should not retry?

Reply:
if we didn't find it above, tried to sync and errored, I don't think there's any point in calling findKeyForKeyID again, right? can just return

Reply:
ohh, sorry! I mis-understood. Will do.

Reply:
Done